### PR TITLE
Did some refactoring, organized into packages

### PR DESCRIPTION
**Pull request by peiva-git**
Let me know what you think. As for #56, I may be wrong, but I think we'll have to implement a new `Game` class to use with the GUI, since we'll be using graphics libraries.

### CODE REVIEW

**Review by manuelkoso — COMMENTED** [review]
I have some doubts about this refactoring:
1. Are the interfaces `Board` and `Cell` and the `FreedomBoard` class necessary? I think that the `Board` class was quite general. 
2. I would put the `FreedomCell` class right at the beginning of the `FreedomBoard` class (same thing for the `Move` class in the `FreedomGame`)
3. Why is the `FreedomCell` static? I'm missing something 
4. The implementation of the `FreedomGame` class still depends on how we get the `userInput`. I don't know but it seems to me "wrong"
5. I have to think about the `Game` interface. We are using the public method `playTurn()` as a private method in the `start()` method. This can be done, but if we use, for example, the `playTurn()` outside the class instead of using the `start()` method, we will don't know when the game will be finished. The public method `playTurn()` will also create some troubles with the public method `getTheWinner()`   
6. Maybe, we have to think about how general we want this thing. Do we consider all types of games with the `Game` interface? What kind of board do we consider? How many players?

**Comment by manuelkoso** [comment]
Instead, we can start with the GUI part to see how to redesign the software architecture

**Comment by peiva-git** [comment]
Starting with the GUI part will definitely help. Some notes:

1. I thought that in general, a `Board` may have more specific methods depending on the used pieces. For example, in our case, we're using black and white `Stone`s, therefore we have the `putStone()` and `getStone()` methods
2. This was done automatically by the refactoring mechanism
3. A `Cell` can exist even when there's no `Board` instance, therefore it's defined as `static`
4. I don't see any way around this, perhaps we'll find one later
5. Very true, `playTurn()` should be `private`
6. I'd say we consider board games only, otherwise it's too vague

File: README.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ From: https://boardgamegeek.com/boardgame/100480/freedom
 
 ### Rules:
```
Comment (manuelkoso) [inline README.md:14]: Pay attention with name refactoring

File: src/main/java/it/units/sdm/project/interfaces/Board.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package it.units.sdm.project.interfaces;
+
+import it.units.sdm.project.Position;
+import org.jetbrains.annotations.NotNull;
+
+
+public interface Board {
```
Comment (manuelkoso) [inline src/main/java/it/units/sdm/project/interfaces/Board.java:7]: You said: *I thought that in general, a Board may have more specific methods depending on the used pieces. For example, in our case, we're using black and white Stones, therefore we have the putStone() and getStone() methods*
I think that the Board and Cell classes are the same for all the games. Instead, we can generalize Stone, so we will have putSomething and getSomething. I would prefer to build an util class, instead, to implement more specific game related methods. The Board must not be dependent on the type of the games.

**Review by manuelkoso — COMMENTED** [review]
I have still some doubts about Board, see the comment. For the moment the Game class can be implemented in that way, let's see after the GUI part.

File: README.md (hunk 1 above)
Comment (peiva-git) [reply README.md:14]: ?

Comment (manuelkoso) [reply README.md:14]: README:
- From: Board: Freedom is played on a 10×10 square board. Beginners can try the game on a 8×8 board. Other board sizes may be also used.
- To: it.units.sdm.project.core.FreedomBoard: Freedom is played on a 10×10 square board. Beginners can try the game on a 8×8 board. Other board sizes may be also used.

Comment (manuelkoso) [reply README.md:14]: I didn't select the proper lines, sorry

File: src/main/java/it/units/sdm/project/interfaces/Board.java (hunk 2 above)
Comment (peiva-git) [reply src/main/java/it/units/sdm/project/interfaces/Board.java:7]: I guess we could do something similar to the `List<T>` java interface, where `T` would be `Stone` in our case

Comment (manuelkoso) [reply src/main/java/it/units/sdm/project/interfaces/Board.java:7]: Yes, great idea!

Comment (peiva-git) [reply src/main/java/it/units/sdm/project/interfaces/Board.java]: I uploaded a first attempt, but as you can see and as I feared there's not much improvement regarding readability and overall class structure. The `FreedomCell` class and `Cell` interface are now basically the same aside from the `stone` status variable

File: README.md [hunk 3]
```diff
@@ -12,7 +12,7 @@ From: https://boardgamegeek.com/boardgame/100480/freedom
...
 
-- **Board**: Freedom is played on a 10×10 square board. Beginners can try the game on a 8×8 board. Other board sizes may be also used.
+- **it.units.sdm.project.core.FreedomBoard**: Freedom is played on a 10×10 square board. Beginners can try the game on a 8×8 board. Other board sizes may be also used.
```
Comment (manuelkoso) [inline README.md:15]: I ment this before.

File: src/main/java/it/units/sdm/project/interfaces/Board.java (hunk 2 above)
Comment (manuelkoso) [reply src/main/java/it/units/sdm/project/interfaces/Board.java]: I see, but as I told you before I don't think that Board has to be an interface (neither Cell). If we are going to implement other games, we have to ovveride the same methods that would have almost the same implementation.
We will talk about it tonight on the meeting I think.

Comment (manuelkoso) [reply src/main/java/it/units/sdm/project/interfaces/Board.java]: > Yes, great idea!

Actually not, lol (at least for me).

**Review by manuelkoso — COMMENTED** [review]
See [comment](https://github.com/peiva-git/SDM_project/pull/57#discussion_r1154276424)

**Comment by manuelkoso** [comment]
*A Cell can exist even when there's no Board instance* 
Are you sure about that? A cell exists whether there is a board.

**Comment by peiva-git** [comment]
I wrote the generic `Board` implementation, as discussed. I encountered an issue with the `toString()` method in the `MapBoard<P>` class, since obviously we can't print generic pieces by assuming they have a color. My solution was this
```java
public class MapBoard<P extends Stone> implements Board<P> {
// etc
}
```
which I think makes sense, considering how a `Stone` represents a very basic piece, having only the `Color`

**Comment by manuelkoso** [comment]
We can override the Stone toString instead!

**Comment by manuelkoso** [comment]
I don't remember, but can we do P.toString()??

**Comment by peiva-git** [comment]
Not instead, we still need `<P extends Stone`, and I'm not sure, since`Stone` is a `public` class. Do we really want each `Stone`'s representation to be only `W` or `B`?

**Comment by manuelkoso** [comment]
> Not instead, we still need `<P extends Stone>`, and I'm not sure, since`Stone` is a `public` class. Do we really want each `Stone`'s representation to be only `W` or `B`?

hmmm... Let's leave it like this and see

**Comment by manuelkoso** [comment]
Last question: if we consider a game that uses red and blue pieces, how would we deal with that?

**Comment by peiva-git** [comment]
We would need to use a `String` color property instead of our `enum Color`

**Comment by manuelkoso** [comment]
> We would need to use a `String` color property instead of our `enum Color`, but I don't think that makes sense, our "color" only represents the presence of two players

hmmmmm... We need to think about it a little more later. 
Ok, I will merge this.

**Review by manuelkoso — APPROVED** [review]
Well done!

**Pull request merged**